client/rpc.go

```diff
@@ -4,70 +4,260 @@ package client
 
 import (
 	"context"
+	"encoding/json"
 	"fmt"
-	"net/rpc"
+	"net/url"
+	"strconv"
+	"sync"
+	"sync/atomic"
 	"time"
 
 	"github.com/ethereum/go-ethereum/common"
+	gorilla "github.com/gorilla/websocket"
 	"github.com/perun-network/erdstall/tee"
-	"github.com/pkg/errors"
+	"github.com/perun-network/erdstall/wire"
+	perunlog "perun.network/go-perun/log"
+	pkgsync "perun.network/go-perun/pkg/sync"
 )
 
-// RPC warps go rpc calls to connect the client with the operator.
-type RPC struct {
-	conn *rpc.Client
-}
+type (
+	// RPC connects the client with the operator over websockets.
+	RPC struct {
+		pkgsync.Closer
+		perunlog.Embedding
+
+		connMtx sync.Mutex // protects conn.send.
+		conn    *gorilla.Conn
+
+		cbMtx     sync.RWMutex // protects callbacks.
+		id        uint64
+		callbacks map[wire.ID]callback
+
+		subMtx       sync.RWMutex // protects subscription.
```

Comment:
Not sure if even needed, see other comment.

Reply:
I had it all stuffed into the `mtx` before, but it deadlocked. Not sure what the cause was though.